SwiftCompilerSources/Sources/Optimizer/FunctionPasses/DeadStoreElimination.swift

```diff
@@ -0,0 +1,267 @@
+//===--- DeadStoreElimination.swift ----------------------------------------==//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+/// Eliminates dead store instructions.
+///
+/// A store is dead if after the store has occurred:
+///
+/// 1. The value in memory is not read until the memory object is deallocated:
+///
+///     %1 = alloc_stack
+///     ...
+///     store %2 to %1
+///     ...               // no reads from %1
+///     dealloc_stack %1
+///
+/// 2. The value in memory is overwritten by another store before any potential read:
+///
+///     store %2 to %1
+///     ...               // no reads from %1
+///     store %3 to %1
+///
+/// In case of a partial dead store, the store is split so that some of the new
+/// individual stores can be eliminated in the next round of the optimization:
+///
+///     store %2 to %1                          // partially dead
+///     ...               // no reads from %1
+///     %3 = struct_element_addr %1, #field1
+///     store %7 to %3
+/// ->
+///     %3 = struct_extract %2, #field1
+///     %4 = struct_element_addr %1, #field1
+///     store %3 to %4                          // this store is dead now
+///     %5 = struct_extract %2, #field2
+///     %6 = struct_element_addr %1, #field2
+///     store %5 to %6
+///     ...              // no reads from %1
+///     store %7 to %3
+///
+let deadStoreElimination = FunctionPass(name: "dead-store-elimination") {
+  (function: Function, context: FunctionPassContext) in
+
+  for block in function.blocks {
+
+    // We cannot use for-in iteration here because if the store is split, the new
+    // individual stores are inserted right afterwards and they would be ignored by a for-in iteration.
```

Comment:
That's how sequences work in swift: the `next()` function (which is called at the begin of the loop body) returns the current element and at advances to the next element in the sequence.
And the SIL linked lists cannot be Collections, because for a stupid reason it's not possible to make linked lists a Collection.